pkgr.yml

```diff
@@ -0,0 +1,28 @@
+Version: 1
+Descriptions:
+  - DESCRIPTION
+
+Packages:
+  - devtools
+
+Repos:
+  - MPN: https://mpn.metworx.com/snapshots/stable/2020-03-24
+  - MPN_Rcpp: https://mpn.metworx.com/snapshots/stable/2020-04-18
+  - MPN_withr: https://mpn.metworx.com/snapshots/stable/2020-05-09
+
+Customizations:
+  Repos:
+    - MPN:
+        Type: source
```

Comment:
why are you forcing source on all of these?

Reply:
Here's the thought process:

1. The default MPN snapshot should be the oldest compatible, precisely to avoid inadvertently relying on new features or backwards-incompatible behavior, as you say. 

2. The current oldest compatible snapshot is 2020-03-24.

3. We want to test against the latest R version for both the current and previous major.minor releases. At the moment, those are 4.0 and 3.6.

4. We test against Linux (Ubuntu) in CI, and we also want to test against macOS, especially given recent platform-specific issues.

5. (macOS) binaries for R 4.0 do not exist for 2020-03-24.

6. pkgr does not fall back to source if binaries are unavailable.

7. Therefore, to have a single `pkgr.yml` that works for both Linux and macOS development, we need to specify `Type: source` for each of the `Repos`.

You also commented that the oldest snapshot should be the "least important" for day-to-day development. Why do you think this? If the concern is that we want to mitigate the risk of (newer versions of) dependencies introducing breaking changes, that should be addressed by the `cran-latest` CI image.

So, the idea would be that development is done against the "oldest compatible" package versions, and if CI reveals a breaking change in a dependency, we will need to decide how to handle, whether by introducing/bumping a version constraint, adding a shim, etc. A similar process would play out if we wanted to take advantage of new, non-breaking functionality in a dependency. The general philosophy is conservative: take the minimal set of version constraints possible.